src/main/java/net/mcreator/element/types/Item.java

```diff
@@ -220,14 +223,105 @@ public List<StateEntry> getModels() {
+	 * For use by 1.21.4+ generators.
+	 *
+	 * @return Models tree with contents matching current generator.
+	 */
+	public StateTreeNode getModelsTree() {
+		StateEntry baseState = getBaseState();
+
+		// initially set to item's default state
+		// if no model overrides are defined, this will return tree of depth 1
+		StateTreeNode root = baseState;
+		for (StateEntry state : getModels()) {
+			if (state.stateMap.isEmpty())
+				continue; // should never happen because states with empty StateMap are removed by getModels()
+
+			StateBranch current = new StateBranch(null, root);
+			Map.Entry<PropertyData<?>, Object> cachedPair = null;
+			for (Map.Entry<PropertyData<?>, Object> pair : state.stateMap.entrySet()) {
+				if (cachedPair != null) {
+					StateTreeNode node = current.values.get(cachedPair.getValue());
+					if (node instanceof StateBranch branch) {
+						current = branch; // switch to found branch to handle it later
+					} else { // otherwise create a new branch for current property
+						current.values.put(cachedPair.getValue(),
+								new StateBranch(pair.getKey(), Objects.requireNonNullElse(node, baseState)));
+						current = (StateBranch) current.values.get(cachedPair.getValue());
+					}
+				}
+
+				// these two will only be called if we are at a wrong branch
+				while (!pair.getKey().equals(current.property) && current.fallback instanceof StateBranch branch)
+					current = branch; // keep going to fallback node until we reach branch for given property
+				if (!pair.getKey().equals(current.property)) { // or create it if not found
+					current.fallback = new StateBranch(pair.getKey(), current.fallback);
+					current = (StateBranch) current.fallback;
+					if (root == baseState) // update reference to root node to be returned
+						root = current; // this will not happen more than once (when the first branch is created)
+				}
+
+				cachedPair = pair;
+			}
+
+			StateTreeNode node = current.values.get(cachedPair.getValue());
+			if (node instanceof StateBranch branch) {
+				// find the deepest fallback node that is a branch
+				current = branch;
+				while (current.fallback instanceof StateBranch branch2)
+					current = branch2;
+				if (current.fallback == null || current.fallback == baseState)
+					current.fallback = state; // only reassign if currently unset or set to default state
+			} else if (node == null || node == baseState) {
+				current.values.put(cachedPair.getValue(), state);
+			}
+		}
+
+		return root;
+	}
+
+	public interface StateTreeNode {
+	}
+
+	public static class StateBranch implements StateTreeNode {
+
+		public PropertyData<?> property;
+		@TextureReference(TextureType.ITEM) @ResourceReference("model")
+		public LinkedHashMap<Object, StateTreeNode> values;
+		@TextureReference(TextureType.ITEM) @ResourceReference("model") public StateTreeNode fallback;
+
+		private StateBranch() {

```

Comment:
Why is this one needed?

Reply:
It just hides the default constructor so that we use the one that initializes all fields for us.

Reply:
But if you specify a custom constructor, the default one is not accessible anyways.